tsdb/mapper.go

```diff
@@ -360,671 +232,364 @@ func (lm *SelectMapper) nextChunkRaw() (interface{}, error) {
-// then an empty slice is returned.
-func (lm *SelectMapper) Fields() []string {
-	if lm.remote != nil {
-		return lm.remote.Fields()
+	// Ensure that the start time for the results is on the start of the window.
+	m.qminWindow = m.qmin
+	if m.intervalSize > 0 && m.intervalN > 1 {
+		m.qminWindow = m.qminWindow / m.intervalSize * m.intervalSize
 	}
-	return append(lm.selectFields, lm.selectTags...)
-}
 
-// Close closes the mapper.
-func (lm *SelectMapper) Close() {
-	if lm.remote != nil {
-		lm.remote.Close()
-		return
-	}
-	if lm != nil && lm.tx != nil {
-		_ = lm.tx.Rollback()
+	// Get a read-only transaction.
+	tx, err := m.shard.engine.Begin(false)
+	if err != nil {
+		return err
 	}
-}
-
-// aggTagSetCursor wraps a standard tagSetCursor, such that the values it emits are aggregated
-// by intervals.
-type aggTagSetCursor struct {
-	nextFunc func() (time int64, value interface{})
-	tagsFunc func() map[string]string
-	tMinFunc func() int64
-}
-
-// Next returns the next value for the aggTagSetCursor. It implements the interface expected
-// by the mapping functions.
-func (a *aggTagSetCursor) Next() (time int64, value interface{}) {
-	return a.nextFunc()
-}
-
-// Tags returns the current tags for the cursor
-func (a *aggTagSetCursor) Tags() map[string]string {
-	return a.tagsFunc()
-}
+	m.tx = tx
 
-// TMin returns the current floor time for the bucket being worked on
-func (a *aggTagSetCursor) TMin() int64 {
-	return a.tMinFunc()
-}
+	// Collect measurements.
+	mms := Measurements(m.shard.index.MeasurementsByName(m.stmt.SourceNames()))
+	m.selectFields = mms.SelectFields(m.stmt)
+	m.selectTags = mms.SelectTags(m.stmt)
+	m.whereFields = mms.WhereFields(m.stmt)
 
-type pointHeapItem struct {
-	timestamp int64
-	value     []byte
-	cursor    *seriesCursor // cursor whence pointHeapItem came
-}
-
-type pointHeap []*pointHeapItem
+	// Open cursors for each measurement.
+	for _, mm := range mms {
+		if err := m.openMeasurement(mm); err != nil {
+			return err
+		}
+	}
 
-func newPointHeap() *pointHeap {
-	q := make(pointHeap, 0)
-	heap.Init(&q)
-	return &q
+	return nil
 }
 
-func (pq pointHeap) Len() int { return len(pq) }
-
-func (pq pointHeap) Less(i, j int) bool {
-	// We want a min-heap (points in chronological order), so use less than.
-	return pq[i].timestamp < pq[j].timestamp
-}
+func (m *AggregateMapper) openMeasurement(mm *Measurement) error {
+	// Validate that ANY GROUP BY is not a field for the measurement.
+	if err := mm.ValidateGroupBy(m.stmt); err != nil {
+		return err
+	}
 
-func (pq pointHeap) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i] }
+	// Validate the fields and tags asked for exist and keep track of which are in the select vs the where
+	selectFields := mm.SelectFields(m.stmt)
+	selectTags := mm.SelectTags(m.stmt)
 
-func (pq *pointHeap) Push(x interface{}) {
-	item := x.(*pointHeapItem)
-	*pq = append(*pq, item)
-}
+	// If we only have tags in our select clause we just return
```

Comment:
nit: this comment doesn't seem to line up with the current behavior.  We actually return an error.